Added dynamic tab alignment support in MultiPageEditorPart

praveen-skp opened the pull request:
Issue: #2223 

This pull request introduces a new feature in Eclipse that allows users to configure the alignment of tabs in multi-page editors. The tabs, which are currently fixed at the bottom, can now be dynamically positioned on the top or bottom, based on user preference. 

Key Changes:
- Added a new preference for multi-page editor tab alignment.
- Added a preference change listener to MultiPageEditorPart.
- Updated tab style based on the user's preference.

New preference: 
![New preference selected](https://github.com/user-attachments/assets/4ca0fbf9-fa2e-4d54-8349-cd949e5ad41c)

Tabs on top of the screen:
![Expected look](https://github.com/user-attachments/assets/7ccbcbd6-a42a-4804-be22-7128eee1cc88)

praveen-skp committed:
Added dynamic tab alignment support in MultiPageEditorPart

- Added a new preference for multi-page editor tab alignment.
- Added a preference change listener to MultiPageEditorPart.
- Updated tab style based on the user's preference.

iloveeclipse commented on bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/part/MultiPageEditorPart.java:
```diff
@@ -291,6 +317,31 @@ protected CTabFolder createContainer(Composite parent) {
+	 *         {@code SWT.BOTTOM} if the user prefers tabs to be aligned on the
+	 *         bottom.
+	 */
+	private int getPreferredTabStyle() {
```
I **believe** this should be made protected and allow concrete editor implementations to override it.

laeubi replied:
```suggestion
	protected int getTabStyle() {
```

iloveeclipse reviewed (commented):
This is an interesting feature, but ideally it should also give custom multipage editors control whether they want to follow up the tab position or keep some specific tab position independently of the preference, see also coment in code.

laeubi commented on bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/part/MultiPageEditorPart.java:
```diff
@@ -291,6 +317,31 @@ protected CTabFolder createContainer(Composite parent) {
+	 */
+	private int getPreferredTabStyle() {
+		boolean alignTabsOnTop = getAPIPreferenceStore()
+				.getBoolean(IWorkbenchPreferenceConstants.ALIGN_MULTI_PAGE_EDITOR_TABS_ON_TOP);
```
Instead of a boolean can't we simply use an INT here?

laeubi commented on bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/part/MultiPageEditorPart.java:
```diff
@@ -291,6 +317,31 @@ protected CTabFolder createContainer(Composite parent) {
+		boolean alignTabsOnTop = getAPIPreferenceStore()
+				.getBoolean(IWorkbenchPreferenceConstants.ALIGN_MULTI_PAGE_EDITOR_TABS_ON_TOP);
+		int style = alignTabsOnTop ? SWT.TOP : SWT.BOTTOM;
+		return style;
```
```suggestion
		return style | SWT.FLAT;
```

laeubi commented on bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/dialogs/EditorsPreferencePage.java:
```diff
@@ -130,6 +132,15 @@ protected void createPromptWhenStillOpenPref(Composite composite) {
 	}
 
+	protected void createAlignMultiPageEditorTabsOnTop(Composite composite) {
+		alignMultiPageEditorTabsOnTop = new Button(composite, SWT.CHECK);
```
I would make it drowndown (or radio group) with TOP / BOTTOM

laeubi commented on bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/part/MultiPageEditorPart.java:
```diff
@@ -149,10 +154,31 @@ public abstract class MultiPageEditorPart extends EditorPart implements IPageCha
+	 * @param event the {@link PropertyChangeEvent} triggered by a change in the
+	 *              preference store
+	 */
+	private void handlePropertyChange(PropertyChangeEvent event) {
```
```suggestion
	protected boolean handlePropertyChange(PropertyChangeEvent event) {
```

I would let this method simpyl return if an update is required and handle the update container in the event listener.

laeubi commented on bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/part/MultiPageEditorPart.java:
```diff
@@ -1230,4 +1281,13 @@ public void run() {
 		}
 	}
+
+	private void updateContainer() {
```
```suggestion
	protected void updateContainer() {
```

tomaswolf commented:
> ... ideally it should also give custom multipage editors control whether they want to follow up the tab position or keep some specific tab position independently of the preference...

Not sure about this. UI consistency is important. If every multipage editor gets to decide on its own, we'll end up with some having the tabs at the top and some at the bottom. Isn't that going to be confusing or at least inconvenient for users?

iloveeclipse commented:
> Isn't that going to be confusing or at least inconvenient for users?

I'm thinking about products like ours where the editors tab position is important because we have lot of complicated and nested UI elements, so users "occasionally" changing global preference would see "broken" UI's in the places where nothing was changed for years.

praveen-skp committed:
Incorporated the review comments

- Changed the check box to drop down with Top and Bottom options
- Made the private methods protected
- Converted the preference store value from boolean to int and storing
the SWT value of the selection directly

praveen-skp committed:
Changed the updateContainer method to protected

praveen-skp commented:
Hi @iloveeclipse, Hi @laeubi ,
I have incorporated the review comments. Could you please go through the change once again ?

BeckerWdf commented:
> New preference: ![New preference selected](https://private-user-images.githubusercontent.com/18184080/361665032-4ca0fbf9-fa2e-4d54-8349-cd949e5ad41c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjY3MjQwNjQsIm5iZiI6MTcyNjcyMzc2NCwicGF0aCI6Ii8xODE4NDA4MC8zNjE2NjUwMzItNGNhMGZiZjktZmEyZS00ZDU0LTgzNDktY2Q5NDllNWFkNDFjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTE5VDA1MjkyNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA2ZWVhMDZlOGZmYjNiMzhjMGMyNzE1ZTQxZTE2NWNkMWRiNDBhNDhiZGE3NTA1M2NlZWEwNGZkZDM2NDg1YTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.p9FmbCsGr33LO7KzVUAc8DrztKXVuz_UF23pHZHpFPw)

I wondering if we should move the new checkbox above the "Number of open editor..." setting so that all checkboxes are next to each other?
Did you move it below the "Number of open editor..." by intention?

praveen-skp commented:
![Dropdown](https://github.com/user-attachments/assets/f008d950-5fe2-404c-b56a-3f23e5f0571a)
Sorry, i forgot to add the current UI after the change. Its now changed from check box to dropdown, after incorporating the review comments.

BeckerWdf commented:
> Sorry, i forgot to add the current UI after the change. Its now changed from check box to dropdown, after incorporating the review comments.

Ah yes. Forgot that. In that case I think the ordering is ok. And I just realised that the "Automatically close.." checkbox and the next input field belong together. One can see this via the indentation. Maybe we could make this more explicit but putting a box around these two - but that would be a different issue.

BeckerWdf commented:
> but that would be a different issue

Created https://github.com/eclipse-platform/eclipse.platform.ui/issues/2297 for this.

BeckerWdf commented:
> I'm thinking about products like ours where the editors tab position is important because we have lot of complicated and nested UI elements, so users "occasionally" changing global preference would see "broken" UI's in the places where nothing was changed for years.

I have some doubts: So with the current state the user could set the preference but some individual editor may not consider this. Isn't this strange behaviour from the user's perspective.
I see your use-case but can we find another solution for it? E.g. having a mechanisms for a product definition to hide this new preference so that in your product the user cannot choose to have editor tabs at the top?

BeckerWdf commented:
@praveen-skp: Can you pls. rebase an resolve the merge conflicts?

BeckerWdf commented:
I see:

`03:17:33.152 [WARNING] MavenProject: org.eclipse.platform:org.eclipse.ui.editors:3.18.100-SNAPSHOT @ /home/jenkins/agent/workspace/eclipse.platform.ui_PR-2224/bundles/org.eclipse.ui.editors/.polyglot.META-INF: baseline and build artifacts have same version but different contents`

Where does this come from?

BeckerWdf commented:
@jukzi:

I see failing test with:

```
Cannot invoke "java.util.function.Supplier.get()" because "this.shell" is null
java.lang.NullPointerException: Cannot invoke "java.util.function.Supplier.get()" because "this.shell" is null
	at org.eclipse.test.Screenshots$ScreenshotOnFailure.dispose(Screenshots.java:52)
	at org.eclipse.test.Screenshots$ScreenshotOnFailure.finished(Screenshots.java:47)
```

Can you pls. have a look what's the issue there?

jukzi commented:
Please consider rebase/merge of my 2 outstanding PRs to use nonnull Argument to get Screenshots

Jörg Kubitz

> Am 14.10.2024 um 07:36 schrieb Matthias Becker ***@***.***>:
> 
> ﻿
> @jukzi:
> 
> I see failing test with:
> 
> Cannot invoke "java.util.function.Supplier.get()" because "this.shell" is null
> java.lang.NullPointerException: Cannot invoke "java.util.function.Supplier.get()" because "this.shell" is null
> 	at org.eclipse.test.Screenshots$ScreenshotOnFailure.dispose(Screenshots.java:52)
> 	at org.eclipse.test.Screenshots$ScreenshotOnFailure.finished(Screenshots.java:47)
> Can you pls. have a look what's the issue there?
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were mentioned.

BeckerWdf commented:
@azoitl: Referring to our talk on OCX.
You said this feature is already available somewhere?

azoitl commented:
> @azoitl: Referring to our talk on OCX. You said this feature is already available somewhere?

No not yet. There is a bugzilla entry: https://bugs.eclipse.org/bugs/show_bug.cgi?id=58945 there you can find links to the gerrit patches I did. But I do not recommend to use them. There is someone who will start working on that.

BeckerWdf commented:
pr-head is failing with

`Failed to execute goal org.eclipse.tycho:tycho-surefire-plugin:4.0.9:test (default-test) on project org.eclipse.ui.tests: An unexpected error occurred while launching the test runtime (process returned error code 13). The process logfile /home/jenkins/agent/workspace/eclipse.platform.ui_PR-2224/tests/org.eclipse.ui.tests/target/work/data/.metadata/.log might contain further details. Command-line used to launch the sub-process was /opt/tools/java/openjdk/jdk-17/latest/bin/java -Dosgi.noShutdown=false -Dosgi.os=linux -Dosgi.ws=gtk -Dosgi.arch=x86_64 --add-modules=ALL-SYSTEM -Dosgi.clean=true -ea -jar /home/jenkins/agent/workspace/eclipse.platform.ui_PR-2224/.m2/repository/p2/osgi/bundle/org.eclipse.equinox.launcher/1.6.900.v20240613-2009/org.eclipse.equinox.launcher-1.6.900.v20240613-2009.jar -data /home/jenkins/agent/workspace/eclipse.platform.ui_PR-2224/tests/org.eclipse.ui.tests/target/work/data -install /home/jenkins/agent/workspace/eclipse.platform.ui_PR-2224/tests/org.eclipse.ui.tests/target/work -configuration /home/jenkins/agent/workspace/eclipse.platform.ui_PR-2224/tests/org.eclipse.ui.tests/target/work/configuration -application org.eclipse.tycho.surefire.osgibooter.uitest -testproperties /home/jenkins/agent/workspace/eclipse.platform.ui_PR-2224/tests/org.eclipse.ui.tests/target/surefire.properties in working directory /home/jenkins/agent/workspace/eclipse.platform.ui_PR-2224/tests/org.eclipse.ui.tests -> [Help 1]`

This looks unrelated. Should we merge it anyways?

jukzi commented:
> This looks unrelated. Should we merge it anyways?

Please make sure the CI will not fail after any commit. for that you can rerun the build , document (with an issue) or even fix the failure

BeckerWdf commented:
failed test on macOS is already know to fail. it's unrelated See: https://github.com/eclipse-platform/eclipse.platform.ui/issues/370.

If nobody objects I plan to merge this today.
